File: competence-database/src/main/scala/uzuzjmd/competence/persistence/dao/LearningProjectTemplate.scala
package uzuzjmd.competence.persistence.dao

import uzuzjmd.competence.persistence.abstractlayer.CompOntologyManager
import uzuzjmd.competence.persistence.ontology.{CompObjectProperties, CompOntClass}

import scala.collection.JavaConverters._

case class LearningProjectTemplate(comp: CompOntologyManager, var name: String, associatedComptences: Seq[Competence] = null, readableName: String = null) extends CompetenceOntologyDao(comp, CompOntClass.LearningProjectTemplate, name) {

  def DEFINITION = "definition"

  @Override
  def getFullDao(): LearningProjectTemplate = {
    return new LearningProjectTemplate(comp, name, getAssociatedCompetences, getDataField(DEFINITION))
  }

  @Override
  def persistMore() {
    if (associatedComptences != null && !associatedComptences.isEmpty) {
      // könnte man jetzt auch die Kompetenzen persistieren
      associatedComptences.foreach(x => addCompetenceToProject(x))
    }

    if (readableName != null) {
      addDataField(DEFINITION, readableName);
    } else {
      addDataField(DEFINITION, name);
    }
  }

  @Override
  def deleteMore() {

  }

  def getAssociatedCompetences(): Seq[Competence] = {
    return getAssociatedSingletonDaosAsRange(CompObjectProperties.LearningProjectTemplateOf, classOf[Competence])
  }

  def getAssociatedCompetencesAsJava(): java.util.List[Competence] = {
    return getAssociatedCompetences.asJava
  }

  def addCompetenceToProject(competence: Competence) {
    createEdgeWith(CompObjectProperties.LearningProjectTemplateOf, competence)
  }

}

File: competence-database/src/main/scala/uzuzjmd/competence/monopersistence/daos/LearningProjectTemplate.java
package uzuzjmd.competence.monopersistence.daos;

import uzuzjmd.competence.persistence.ontology.CompObjectProperties;
import uzuzjmd.competence.shared.dto.LearningTemplateResultSet;

import java.util.List;

/**
 * Created by dehne on 11.01.2016.
 */
public class LearningProjectTemplate extends AbstractLearningProjectTemplate implements HasDefinition, Cascadable{

    public LearningProjectTemplate(String id) {
        super(id);
    }

    public LearningProjectTemplate(String id, List<Competence> associatedCompetences) {
        super(id);
        this.associatedCompetences = associatedCompetences;
    }





    /**
     * Definition here is the same as the readable Name
     * @return
     */
    @Override
    public String getDefinition() {
        return this.getId();
    }

    @Override
    public void persistMore() throws Exception {
        this.persist();
        if (associatedCompetences != null) {
            for (Competence associatedCompetence : associatedCompetences) {
                addCompetenceToProject(associatedCompetence);
            }
        }
    }


    public List<Competence> getAssociatedCompetences() throws Exception {
        return getAssociatedDaosAsDomain(CompObjectProperties.LearningProjectTemplateOf, Competence.class);
    }


    public void addCompetenceToProject(Competence competence) throws Exception {
        createEdgeWith(CompObjectProperties.LearningProjectTemplateOf, competence);
    }
}
